### PR TITLE
feat: support StarCoder model architectures

### PR DESCRIPTION
**Pull request by wsxiaoys**
#3076 

<s>Still in progress, while the model converting / params loading part seems to be working.</s>

[Tabby](https://github.com/TabbyML/tabby) has integrated llama.cpp and released v0.1.1 🎉. It now offers native support for metal inference and the StarCoder model!

### CODE REVIEW

**Comment by ggerganov** [comment]
Looks good so far - let us know if you hit any roadblocks

**Comment by wsxiaoys** [comment]
> Looks good so far - let us know if you hit any roadblocks

The remaining part for now is from line 3580 to line 3718 in llama.cpp. It should not be very hard to figure it out once I have set up a development environment to ensure the matrix shape arithmetic is correct...

**Comment by wsxiaoys** [comment]
OK, I think I got a version running under CPU:

```bash
> make main && ./bin/main -m ../models/starcoder-1b.gguf -p "# Dijkstra's shortest path algorithm in Python (4 spaces indentation) + complexity analysis:\n\n" -e -t 4 --temp -1 -n 128 -ngl 0

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 |
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = -1.000000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 128, n_keep = 0


# Dijkstra's shortest path algorithm in Python (4 spaces indentation) + complexity analysis:


def dijkstra(graph, start):
    """
    Returns the shortest path from `start` to all other nodes in `graph`.

    The graph is represented as a dictionary of dictionaries. Each key represents a node and each value is another dictionary with keys 'to' and 'cost'.
    """
    # Initialize the distances array to infinity
    distances = [float('inf') for _ in range(len(graph))]
    distances[start] = 0

    # Initialize the previous array to None
    previous = [None for _ in range(len(graph))]

    # Loop through all nodes and find the shortest path
llama_print_timings:        load time =   110.20 ms
llama_print_timings:      sample time =   134.80 ms /   128 runs   (    1.05 ms per token,   949.55 tokens per second)
llama_print_timings: prompt eval time =   262.29 ms /    20 tokens (   13.11 ms per token,    76.25 tokens per second)
llama_print_timings:        eval time =  3485.94 ms /   127 runs   (   27.45 ms per token,    36.43 tokens per second)
llama_print_timings:       total time =  3914.92 ms
```

But it's currently buggy in metal:
```bash
> make main && ./bin/main -m ../models/starcoder-1b.gguf -p "# Dijkstra's shortest path algorithm in Python (4 spaces indentation) + complexity analysis:\n\n" -e -t 4 --temp -1 -n 128 -ngl 1


system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 |
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = -1.000000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 128, n_keep = 0


# Dijkstra's shortest path algorithm in Python (4 spaces indentation) + complexity analysis:

<|endoftext|> [end of text]

llama_print_timings:        load time =   232.01 ms
llama_print_timings:      sample time =     1.26 ms /     1 runs   (    1.26 ms per token,   791.14 tokens per second)
llama_print_timings: prompt eval time =    21.64 ms /    20 tokens (    1.08 ms per token,   924.17 tokens per second)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =    23.29 ms
ggml_metal_free: deallocating
Log end
```

Looking into it...

**Comment by alitariq4589** [comment]
Can one of the admins verify this patch?

**Comment by ggerganov** [comment]
@wsxiaoys There was a bug in the soft max Metal kernel. Can you give me access to push a fix?

```bash
 $ ▶ git push tabbyml HEAD:support-starcoder 
remote: Permission to TabbyML/llama.cpp.git denied to ggerganov.
fatal: unable to access 'https://github.com/TabbyML/llama.cpp/': The requested URL returned error: 403
```

Or I can push it to a branch in this repo? Anyway work for me

**Comment by wsxiaoys** [comment]
Thanks for the fix! Will cleanup  the impl a bit then send it out for review

**Comment by wsxiaoys** [comment]
- [ ] Actively implementing MQA (I will work on this soon).
- [ ] Implementing CUDA offloading (I will skip this part and leave it to the community).

**Comment by wsxiaoys** [comment]
PR Ready for review now 🥇

File: llama.cpp [hunk 1]
```diff
@@ -235,6 +238,7 @@ static std::map<llm_kv, std::string> LLM_KV_NAMES = {
     { LLM_KV_FEED_FORWARD_LENGTH,           "%s.feed_forward_length"   },
     { LLM_KV_USE_PARALLEL_RESIDUAL,         "%s.use_parallel_residual" },
     { LLM_KV_TENSOR_DATA_LAYOUT,            "%s.tensor_data_layout"    },
+    { LLM_KV_MAX_POSITION_EMBEDDINGS,       "%s.max_position_embeddings"    },
```
Comment (Green-Sky) [inline llama.cpp:241]: by the name of this, it should be duplicate of `LLM_KV_CONTEXT_LENGTH`

Comment (wsxiaoys) [reply llama.cpp]: Fixed, removed. PTAL

File: llama.cpp [hunk 2]
```diff
@@ -1593,9 +1622,11 @@ std::string llama_model_ftype_name(enum llama_ftype ftype) {
 
 static const char * llama_model_type_name(e_model type) {
     switch (type) {
+        case MODEL_1B:  return "3B";
```
Comment (ggerganov) [inline llama.cpp:1625]:
```suggestion
        case MODEL_1B:  return "1B";
```

File: llama.cpp [hunk 3]
```diff
@@ -3305,6 +3426,243 @@ static struct ggml_cgraph * llm_build_falcon(
...
+
+                ggml_build_forward_expand(gf, ggml_cpy(ctx0, Kcur, k));
+                ggml_build_forward_expand(gf, ggml_cpy(ctx0, Vcur, v));
+            }
+
+            struct ggml_tensor * Q =
+                ggml_permute(ctx0,
+                        ggml_cpy(ctx0,
+                            Qcur,
+                            ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, n_embd/n_head, n_head, N)),
```
Comment (ggerganov) [inline llama.cpp:3562]:
```suggestion
                            ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, n_embd_head, n_head, N)),
```

File: gguf-py/gguf/gguf.py
Comment (ggerganov) [inline gguf-py/gguf/gguf.py:1]: @monatis Do we need to bump `gguf.py` version after this change?

**Review by ggerganov — APPROVED** [review]

File: llama.cpp [hunk 4]
```diff
@@ -3305,6 +3426,243 @@ static struct ggml_cgraph * llm_build_falcon(
...
+#if 1
+            struct ggml_tensor * KQV = ggml_mul_mat(ctx0, V, KQ_soft_max);
+            ggml_set_name(KQV, "KQV");
+#else
+            // make V contiguous in memory to speed up the matmul, however we waste time on the copy
+            // on M1 this is faster for the perplexity computation, but ~5% slower for the single-token generation
+            // is there a better way?
+            struct ggml_tensor * V_cont = ggml_cpy(ctx0, V, ggml_new_tensor_3d(ctx0, kv_self.v->type, n_past + N, n_embd_head, n_head));
+            struct ggml_tensor * KQV = ggml_mul_mat(ctx0, V_cont, KQ_soft_max);
+#endif
```
Comment (ggerganov) [inline llama.cpp:3609]: I shouldn't have copy-pasted this again

```suggestion
            struct ggml_tensor * KQV = ggml_mul_mat(ctx0, V, KQ_soft_max);
            ggml_set_name(KQV, "KQV");
```

**Pull request merged**